File: examples/04_application/main.php
<?php
// Lean Code — BSD 3-Clause License — Vivian Voss, 2026
// Scope: Entry point. Process all orders into invoices.

require_once __DIR__ . '/configuration/configuration.php';
require_once __DIR__ . '/order/order.php';
require_once __DIR__ . '/stock/stock.php';
require_once __DIR__ . '/billing/billing.php';

$configuration = Configuration\readConfigurationFromFile('config.txt');
$pricePerUnit = (float) $configuration['price_per_unit'];

$orderFiles = Order\findOrderFilesInDirectory($configuration['order_directory']);

foreach ($orderFiles as $file) {
    $order = Order\readOrderFromFile($file);

    if (!Stock\checkStockAvailability($order['product'], $order['quantity'], $configuration['stock_file'])) {
        fwrite(STDERR, "out of stock: {$order['product']}\n");
        continue;
    }

    $invoice = Billing\createInvoiceFromOrder($order, $pricePerUnit);
    Billing\writeInvoiceToFile($invoice, "invoice_{$order['id']}.txt");
}

File: examples/04_application/main.py
# Lean Code — BSD 3-Clause License — Vivian Voss, 2026
# Scope: Entry point. Process all orders into invoices.

import sys
from configuration.configuration import read_configuration_from_file
from order.order import read_order_from_file, find_order_files_in_directory
from stock.stock import check_stock_availability
from billing.billing import create_invoice_from_order, write_invoice_to_file

configuration = read_configuration_from_file("config.txt")
price_per_unit = float(configuration["price_per_unit"])

order_files = find_order_files_in_directory(configuration["order_directory"])

for file in order_files:
    order = read_order_from_file(file)

    if not check_stock_availability(order["product"], order["quantity"], configuration["stock_file"]):
        print(f"out of stock: {order['product']}", file=sys.stderr)
        continue

    invoice = create_invoice_from_order(order, price_per_unit)
    write_invoice_to_file(invoice, f"invoice_{order['id']}.txt")
